crates/bevy_ecs/src/schedule/stage.rs

```diff
@@ -199,51 +227,102 @@ impl SystemStage {
-                match insertion_point {
-                    InsertionPoint::AtStart => {
-                        let index = self.exclusive_at_start.len();
-                        self.uninitialized_at_start.push(index);
-                        self.exclusive_at_start.push(container);
-                    }
-                    InsertionPoint::BeforeCommands => {
-                        let index = self.exclusive_before_commands.len();
-                        self.uninitialized_before_commands.push(index);
-                        self.exclusive_before_commands.push(container);
-                    }
-                    InsertionPoint::AtEnd => {
-                        let index = self.exclusive_at_end.len();
-                        self.uninitialized_at_end.push(index);
-                        self.exclusive_at_end.push(container);
-                    }
-                }
-            }
-            SystemDescriptor::Parallel(descriptor) => {
-                self.uninitialized_parallel.push(self.parallel.len());
-                self.parallel
-                    .push(ParallelSystemContainer::from_descriptor(descriptor, set));
-            }
+        let (run_criteria, mut systems) = system_set.bake();
+        let set_run_criteria_label =
+            run_criteria.map(|criteria| self.process_run_criteria(criteria));
+        for system in systems.drain(..) {
+            self.add_system_inner(system, set_run_criteria_label.clone());
         }
         self
     }
 
+    pub fn with_stage_run_criteria<S: System<In = (), Out = ShouldRun>>(
+        mut self,
+        system: S,
+    ) -> Self {
+        self.set_stage_run_criteria(system);
+        self
+    }
+
+    pub fn set_stage_run_criteria<S: System<In = (), Out = ShouldRun>>(
+        &mut self,
+        system: S,
+    ) -> &mut Self {
+        self.stage_run_criteria.set(Box::new(system));
+        self
+    }
+
+    pub fn with_run_criteria(mut self, run_criteria: RunCriteriaDescriptor) -> Self {
+        self.add_run_criteria(run_criteria);
+        self
+    }
+
+    pub fn add_run_criteria(&mut self, run_criteria: RunCriteriaDescriptor) -> &mut Self {
+        self.process_run_criteria(RunCriteriaDescriptorOrLabel::Descriptor(run_criteria));
+        self
+    }
+
+    fn process_run_criteria(
+        &mut self,
+        run_criteria: RunCriteriaDescriptorOrLabel,
+    ) -> BoxedRunCriteriaLabel {
+        let mut descriptor = match run_criteria {
+            RunCriteriaDescriptorOrLabel::Label(label) => return label,
+            RunCriteriaDescriptorOrLabel::Descriptor(descriptor) => descriptor,
+        };
+        let index = self.run_criteria.len();
+        self.uninitialized_run_criteria
+            .push((index, descriptor.duplicate_label_strategy));
+        let label = descriptor
+            .label
+            .take()
+            .unwrap_or_else(|| Box::new(RunCriteriaIndexLabel(index)));
```

Comment:
This is the pre-filtered pre-sorted run criteria list index. I think this has two implications:

1. When duplicates are filtered out, it means that the next time a run criteria is added, we could get a "duplicate" label despite the fact that it is a "new" criteria.
2. The index stored in the label is meaningless (not a huge deal if (1) is fixed and this is called out in docs)